### PR TITLE
Sync taiko 2024-03-27

### PR DESCRIPTION
**Pull request by CeciliaZ030**
- [x] squashed the previous taiko changes from https://github.com/taikoxyz/revm/pull/2
- [x] reapply to the latest upstream
- [x] change `"optimism"` flags to `"flags"` while reserving `"taiko-default-handler"` for BBR
- [x] keep a dummy `L1BlockInfo` for potential usage in accessing L1 state

** Note: looks like some newly included dep breaks `cargo check --target riscv32imac-unknown-none-elf --no-default-features` again so powdr would not compile, but who cares. All vm should support std.

### CODE REVIEW

File: crates/revm/src/taiko/handler_register.rs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+use alloc::sync::Arc;
+use SpecId::{CANCUN, LONDON};
+
+pub fn taiko_handle_register<DB: Database, EXT>(handler: &mut EvmHandler<'_, EXT, DB>) {
+    spec_to_generic!(handler.cfg.spec_id, {
+        handler.pre_execution.deduct_caller = Arc::new(deduct_caller::<SPEC, EXT, DB>);
+        handler.post_execution.reimburse_caller = Arc::new(reimburse_caller::<SPEC, EXT, DB>);
+        handler.post_execution.reward_beneficiary = Arc::new(reward_beneficiary::<SPEC, EXT, DB>);
+        // Done with flags to avoid repetitive code
+        // handler.validation.tx_against_state = Arc::new(mainnet::validate_tx_against_state::<SPEC, EXT, DB>);
```
Comment (Brechtpd) [inline crates/revm/src/taiko/handler_register.rs:19]: remove?

File: crates/revm/src/taiko/handler_register.rs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        .inner
+        .journaled_state
+        .load_account(caller, &mut context.evm.inner.db)?;
+
+    caller_account.info.balance = caller_account
+        .info
+        .balance
+        .saturating_add(effective_gas_price * U256::from(gas.remaining() + gas.refunded() as u64));
+
+    Ok(())
```
Comment (Brechtpd) [inline crates/revm/src/taiko/handler_register.rs:46]:
```suggestion
    mainnet_reimburse_caller::<SPEC, EXT, DB>(context, gas)
```

If you import it as `use crate::handler::mainnet::reimburse_caller as mainnet_reimburse_caller;` looks like we can just call the standard ethereum handler after skipping the anchor tx.

File: crates/revm/src/taiko/handler_register.rs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    let (coinbase_account, _) = context
+        .evm
+        .inner
+        .journaled_state
+        .load_account(beneficiary, &mut context.evm.inner.db)?;
+    coinbase_account.mark_touch();
+    coinbase_account.info.balance = coinbase_account
+        .info
+        .balance
+        .saturating_add(coinbase_gas_price * U256::from(gas.spent() - gas.refunded() as u64));
```
Comment (Brechtpd) [inline crates/revm/src/taiko/handler_register.rs:78]:
```suggestion
    mainnet_reward_beneficiary::<SPEC, EXT, DB>(context, gas)
```
same thing here with import `use crate::handler::mainnet::reward_beneficiary as mainnet_reward_beneficiary;`

File: crates/primitives/src/env.rs [hunk 4]
```diff
@@ -560,27 +573,6 @@ impl TxEnv {
...
-            nonce: None,
-            access_list: Vec::new(),
-            blob_hashes: Vec::new(),
-            max_fee_per_blob_gas: None,
-            #[cfg(feature = "optimism")]
-            optimism: OptimismFields::default(),
-        }
-    }
-}
-
```
[the remark this answers is not in the document]
Comment (CeciliaZ030) [reply crates/primitives/src/env.rs:583]: looks like u have OCD

File: crates/primitives/src/result.rs [hunk 5]
```diff
@@ -179,12 +179,6 @@ impl<DBError> From<InvalidTransaction> for EVMError<DBError> {
     }
 }
 
-impl<DBError> From<InvalidHeader> for EVMError<DBError> {
-    fn from(value: InvalidHeader) -> Self {
-        Self::Header(value)
-    }
-}
-
```
[the remark this answers is not in the document]
Comment (CeciliaZ030) [reply crates/primitives/src/result.rs:187]: So much OCD

File: crates/primitives/src/result.rs [hunk 6]
```diff
@@ -329,28 +295,24 @@ impl fmt::Display for InvalidTransaction {
...
-            #[cfg(feature = "optimism")]
-            Self::HaltedDepositPostRegolith => {
-                write!(
-                    f,
-                    "deposit transaction halted post-regolith; error will be bubbled up to main return handler"
-                )
+            InvalidTransaction::EmptyBlobs => write!(f, "Empty blobs"),
+            InvalidTransaction::BlobCreateTransaction => write!(f, "Blob create transaction"),
+            InvalidTransaction::TooManyBlobs => write!(f, "Too many blobs"),
+            InvalidTransaction::BlobVersionNotSupported => write!(f, "Blob version not supported"),
```
[the remark this answers is not in the document]
Comment (CeciliaZ030) [reply crates/primitives/src/result.rs:301]: Probably from squash commit, will remove

**Pull request merged**